kore/src/Kore/Attribute/Synthetic.hs

```diff
@@ -6,45 +6,104 @@ License     : NCSA
                  ( CofreeF (..) )
 import qualified Control.Comonad.Trans.Cofree as Cofree
+import           Data.Functor.Const
 import           Data.Functor.Foldable
                  ( Base, Corecursive, Recursive )
 import qualified Data.Functor.Foldable as Recursive
 
+{- | @Synthetic@ is the class of synthetic attribute types @syn@.
+
+@Synthetic base syn@ allows synthesizing @syn@ given a @'Cofree' base@ tree;
+that is, a 'Cofree' tree with branching described by a @'Functor' base@.
+
+ -}
+class Functor base => Synthetic base syn where
+    -- | @synthetic@ is the @base@-algebra for synthesizing the attribute @syn@.
+    synthetic :: base syn -> syn
+
+instance Synthetic (Const a) a where
+    synthetic (Const a) = a
+    {-# INLINE synthetic #-}
+
 {- | @/synthesize/@ attribute @b@ bottom-up along a tree @s@.
 
 @synthesize@ is a generalization of 'Data.List.scanr' to trees: Given a tree @s@
-with attributes @a@ along the nodes, @synthesize@ produces a tree @t@ with
-attributes @b@ along the nodes using the given @(Base s)@-algebra from the
-bottom up. The algebra's argument is the original @a@-attribute of a node and
-the @b@-attributes of all children.
+with attributes @inh@ along the nodes, @synthesize@ produces a tree @t@ with
+attributes @syn@ along the nodes using the given @('Base' s)@-algebra from the
+bottom up.
 
 See also:
 <https://en.wikipedia.org/wiki/Attribute_grammar#Synthesized_attributes>
 
  -}
-synthesize
+resynthesize
+    ::  ( Recursive s
+        , Corecursive t
+        , Recursive t
+        , Base s ~ CofreeF base inh
+        , Base t ~ CofreeF base syn
+        , Synthetic base syn
+        )
+    => s  -- ^ Original tree with attributes @a@
+    -> t
+resynthesize = resynthesizeAux synthetic
+{-# INLINE resynthesize #-}
+
+{- | @/resynthesize/@ attribute @b@ bottom-up along a tree @s@.
+
+@resynthesize@ is a generalization of 'Data.List.scanr' to trees: Given a tree
+@s@, @resynthesize@ produces a tree @t@ with attributes @b@ along the nodes
+using the given @(Base s)@-algebra from the bottom up.
+
+See also:
+<https://en.wikipedia.org/wiki/Attribute_grammar#Synthesized_attributes>
+
+ -}
+resynthesizeAux
     ::  ( Functor f
         , Recursive s
         , Corecursive t
         , Recursive t
         , Base s ~ CofreeF f a
         , Base t ~ CofreeF f b
         )
-    => (CofreeF f a b -> b)  -- ^ @(Base s)@-algebra synthesizing @b@
+    => (f b -> b)  -- ^ @(Base s)@-algebra synthesizing @b@
     -> s  -- ^ Original tree with attributes @a@
     -> t
-synthesize synthetic =
-    Recursive.fold synthesize1
+resynthesizeAux synth =
+    Recursive.fold worker
   where
-    synthesize1 (a :< ft) =
-        Recursive.embed (synthetic (a :< fb) :< ft)
-      where
-        fb = Cofree.headF . Recursive.project <$> ft
+    worker (_ :< ft) = synthesizeAux synth ft
+{-# INLINE resynthesizeAux #-}
 
+{- | @/synthesize/@ an attribute @a@ bottom-up along a tree @s@.
```

Comment:
Is this really bottom-up, or does it do stuff only at the top?

Reply:
Ah, you're right, it only does things at the top. I think I copied this from `resynthesize`, which calls `synthesize` at every level.